### ISSUE TITLE
[Baka-Tsuki] Partial epubs

### ISSUE THREAD

**Issue by typhoon71** [dteviot/WebToEpub#98]
I noticed there's an issue asking for a "multi-volume parser" for Baka-Tsuki (Issue #95).

I was thinking of asking for the ability to grab partially translated novels, for example ones where say only illustrations, preface and 3 chapters are translated (and say last 2 chapters and afterword are not available yet).

Doing something by hand (launching webtoepub from another site and pasting links from Baka-Tsuki is suboptimal right now.

**Comment by dteviot** [comment]
@typhoon71
Can you describe what your "ideal solution" for this would look like please.
e.g. What would the UI look like?
What would the user do on the UI.
What would the the resulting actions.
etc.

Could you perhaps write it as if you were giving a set of "How To" instructions for someone to use the feature?

**Comment by typhoon71** [comment]
OK, let's make an example. I imagine it working like this:
 
1) the user goes in the main page of a LN in BakaTsuki: 
https://www.baka-tsuki.org/project/index.php?title=Shinmai_Maou_no_Keiyakusha
2) the user clicks the web to epub icon as normal

3) the extension detects BakaTsuki, and populates the chapter list with all the links in that page, both the fulltext links and the single chapters or gallery links, like:
https://www.baka-tsuki.org/project/index.php?title=Shinmai_Maou_no_Keiyakusha:Volume_7
https://www.baka-tsuki.org/project/index.php?title=Shinmai_Maou_no_Keiyakusha: Volume_7_Illustrations, https://www.baka-tsuki.org/project/index.php?title=Shinmai_Maou_no_Keiyakusha:Volume_7_Prologue
https://www.baka-tsuki.org/project/index.php?title=Shinmai_Maou_no_Keiyakusha:Volume_7_Chapter_1, ...
...
(this would provide a choice for cases where the fulltext link is available but there's a missing afterword, like in volume 6)

5) if the user selects a single fulltext link, the usual BakaTsuki parser is used, maybe just using the existing code, relaunching the extension with only that link as a basis; everything should work as it does now.

6) if the user selects more than one link, a new parser will be launched, and it will check if the links are in the same volume and pack them according, getting the cover image from the pic on the side or letting one choose one from the gallery if available. It should compile the metadata forms based on BakaTsuki naming schema too.
7) the partial volume grab would contain the list of fetched chapters, for future updates (if implemented); the full volume could contain a "code" of sorts to tell web to epup that it's complete.

I explained it as if there were a pre-parser and two "per-case parsers", but it is just as I thought it could work, it doesn't have to be like that, it's only for making it easier for me to explain.

**Comment by dteviot** [comment]
@typhoon71  
Firstly, thank you for the detailed post.  It makes it easier to understand what is required.

As you may have noticed, there have been multiple requests to get WebToEpub (WTE) to work from a "Series Page", rather than a "Full Text" page.  

The first problem I encountered for doing that was how WTE can distinguish between the two page types.  Refer. https://github.com/dteviot/WebToEpub/issues/37.  Note that no solution had been found.  However, I'm now thinking that in **most** cases the URL for a full text page contains a colon, A series page does not.  e.g.  This is a full page URL.
```
https://www.baka-tsuki.org/project/index.php?title=Shinmai_Maou_no_Keiyakusha

This is a series page URL
```
https://www.baka-tsuki.org/project/index.php?title=Shinmai_Maou_no_Keiyakusha:Volume_1

I'm thinking the solution is be for WebToEpub to look at URL and configure itself appropriately, but add a Drop Down list control on the page to allow user to change between "Full Page" and "Series Page" modes in case it's wrong.

@toshiya44 Does this seem reasonable, or do you have a better suggestion?

**Comment by toshiya44** [comment]
I think you can safely use the existence of at least one colon to determine if a page is a **sub** page of a project. 

There may be a few exceptions such as this url https://www.baka-tsuki.org/project/index.php?title=HorizonːKimitoasamade:Volume_1B

Which contains two kinds of colons. One is `U+003A` colon and the other is `U+02D0` "MODIFIER LETTER TRIANGULAR COLON".

I don't know if this will cause any issues when you're parsing the links...

-------------

> 3. [...] (this would provide a choice for cases where the fulltext link is available but there's a missing afterword, like in volume 6)

I don't think it would possible/worth to implement this part. Because, chapter names are not always consistent (i.e. Project:ChaperX) and can vary wildly. It gets even more complicated when there are short stories and such are linked in there, not to mention that a translator can always add extra parts later on. 

In short, it is hard to determine whether a "Volume" is actually complete or not.

So, the extension should simply populate the list of links with all the non-external links in the project page and let the user choose which links to include. (If possible, links leading to third party sites should be followed only if there's a parser available.)

About selecting full text page from list of links: Same issue as chapter names. The name of a full text page is not consistent throughout the site. Different projects use different conventions and languages.

**Comment by dteviot** [comment]
@typhoon71 
Intial work committed to the Experimental Tab branch, if you'd like to give it a try.
How it works:
1. Browse to either a "Full text" or "Series" page.
2. Click on the "Web to Epub" icon.
3. If "Full text" behaves as it has done in the past.
4. If "Series" page, gives the URLs found on page for user to select.  ("Redlined" links and links on same page are not on the list shown.)

Notes:
1. Series mode collects and packs each link individually.  This means that hyperlinks between pages (e.g. Translator notes on separate web pages from text referred to) are not fixed up.
2. Duplicate images are (probably) not culled from the Gallery (I haven't tested to confirm.)  (Same reason as above.)
3. I've added a thumbnail image of the Cover URL beside the cover URL input box.  (Note it only updates if you press "enter", or click on a different control after changing the URL text.)

**Comment by typhoon71** [comment]
Beautiful! Thanks for adding this. Test results:

**GOOD THINGS**
- It works well with The_Unexplored_Summon_Blood_Sign and the images are correctly deduped too.
- It's able to grab chapters that are on different web domains, like Tsukumodo_Antique_Shop vol 5.
- The fulltext grabbing is continuing to work correctly like in the past.

**ISSUES**
- I found some a problem with Toaru_Majutsu_no_Index:_New_Testament, where no chapters are shown, only  the covers. Grabbing that creates a epub with the source page instead, with all the volumes and sections as a chapters. Maybe it's caused by the title format of the novel?
- In "chapters mode", if you choose one of the suggested cover images, it will use a small image and won't browse to the fullres one; the problem is that in this case you can't just cut the resizing bit after the image, but instead browse to the link in the image and grab the fullres into that page.

**Comment by dteviot** [comment]
typhoon71 
> I found some a problem with Toaru_Majutsu_no_Index:_New_Testament, where no chapters are shown, only the covers.

That's because there's a colon in the URL, so it's picking the "Full Page" parser.  When this happens, use the "Manually Select Parser" option under advanced settings to select the correct parser.  Note there are two Baka-Tsuki options now, Series Page and Full Page.

> In "chapters mode", if you choose one of the suggested cover images, it will use a small image

I'll see what I can do.  Until I do, I suggest just put in the URL of the full size image.

**Comment by typhoon71** [comment]
Oh, so there are 2 parsers for BakaTsuki. Noted, tested, it's really handy. Thanks.

For the cover image I'll do what you suggest, i already did actually, but wanted to let you know since it was a bit of a special case.

**Comment by dteviot** [comment]
@typhoon71
> I found some a problem with Toaru_Majutsu_no_Index:_New_Testament, where no chapters are shown, only the covers.

Looking at folowing pages 
* https://www.baka-tsuki.org/project/index.php?title=Category:Light_novel_(English), 
* https://www.baka-tsuki.org/project/index.php?title=Category:Teaser_(English)

there are a number of series pages with a colon in the URL (this doesn't count the non-English projects.)

* https://www.baka-tsuki.org/project/index.php?title=Anohana:_The_Flower_We_Saw_That_Day
* https://www.baka-tsuki.org/project/index.php?title=Dawnbringer:_The_Story_of_the_Machine_God
* https://www.baka-tsuki.org/project/index.php?title=Toaru_Majutsu_no_Index:_New_Testament
* https://www.baka-tsuki.org/project/index.php?title=Houkago_no_Pleiades:_Minato_no_Hoshizora
* https://www.baka-tsuki.org/project/index.php?title=Pocket_Monsters:_The_Animation
* https://www.baka-tsuki.org/project/index.php?title=Steins;Gate:_Butterfly_Effect%27s_Divergence

**Comment by typhoon71** [comment]
well, as long we can select the right parser I'm fine.
If I understand it correctly only the auto detection is affected by the ":" right?

**Comment by dteviot** [comment]
Colon in URL is only used to select between the two Baka-Tsuki parsers.

**Comment by dteviot** [comment]
@typhoon71 @toshiya44
On reflection, I'm not satisfied with current logic to pick between the Baka-Tsuki parsers.
My concern is that someone is going to use one of the pages that doesn't work, get the wrong parser, not know how to fix it by manually selecting the correct parser (You didn't until I told you). And then get upset becuase something that worked before is now broken.
I think the best solution for time being is to remove the "series page" parser from the automatic selection, so the "Full Page" parser will be chosen by default for B-T.
But, allow the Series Page to be manually selected via the "Advanced Options".
This does have the disadvantage that the feature isn't likely to be discovered.
It may be possible to fix that by writing a Help/FAQ document.   Which I should probably do anyway.

Other possibilities:
* Have "Allow selecting 'Series Page' parser" as an advanced option.  With warning that note that if selected incorrectly, can manually change between parsers.  (Although that is going to be a long label for the control.)
*  Give a warning message each time the "Series Page" parser is selected.  But that's going to get annoying.
* Hard-code the exceptions into the selection logic.  Problem with that is: new pages could be added.  I'd need some way to keep it up to date.

Do either of you have any ideas?

**Comment by typhoon71** [comment]
+1 to "Allow selecting 'Series Page' parser" as an advanced option, disabled by default, with warning when enables (first time only)

**Comment by typhoon71** [comment]
I click without thinking... sorry.

**Comment by dteviot** [comment]
>I clicked without thinking... sorry.

No problem.  I've done that myself.

> Have "Allow selecting 'Series Page' parser" as an advanced option. With warning that note that if selected incorrectly, can manually change between parsers. (Although that is going to be a long label for the control.)

I should have said,
"Automatically select 'Baka-Tsuki Series Page' parser" as an advanced option.  With a "more details" hyperlink beside it which when clicked goes to a help page.   (This implies the need for a Help page. Refer https://github.com/dteviot/WebToEpub/issues/147)

**Comment by dteviot** [comment]
@typhoon71 
Have committed "Automatic parser select includes Baka-Tsuki Series Page Parser" to advanced options.  (Default value is off.)
Unfortunately, BT is currently off-line so I'm unable to verify it actually works.

**Comment by typhoon71** [comment]
BT is now online, I tested it and it works OK. Thanks.

**Comment by dteviot** [comment]
@typhoon71 @toshiya44
Can I close this?  I was concerned there might be issues with hyperlinks to translator notes between pages.  However, every story I checked had the translator notes on same web page as the chapter.  So links are correctly fixed up.  Have either of you seen any stories where this is a problem?

**Comment by toshiya44** [comment]
There are a few projects that has notes on a separate page. Would it be possible to try to fetch them?

https://www.baka-tsuki.org/project/index.php?title=Fate/Zero  
https://www.baka-tsuki.org/project/index.php?title=Mushi_to_Medama

**Comment by dteviot** [comment]
@toshiya44
Thanks, I'll see what I can do.

**Comment by dteviot** [comment]
@toshiya44
New commit to experimental tab branch, basic support for changing URLs 
Now, If WebToEpub now collects more than one web page and a page has links to the other, then the links are changed to point to corresponding position in the new EPUB.
I've only done a basic test.  
1. Went to Fate/Zero page with all volumes and chapters. https://www.baka-tsuki.org/project/index.php?title=Fate/Zero, 
2. selected "8 years ago" from Prologue (https://www.baka-tsuki.org/project/index.php?title=Fate/Zero:Prologue_1) and "Translators' Notes and References" (https://www.baka-tsuki.org/project/index.php?title=Fate/Zero:Translator%27s_Notes)
3. Hit "Pack EPUB"

But it seems to work.

Should also work for https://www.baka-tsuki.org/project/index.php?title=Mushi_to_Medama if you go to the Volumes page and include the "Translators' Notes and References" in your slection.

Please let me know if you have any issues.

**Comment by toshiya44** [comment]
It seems that the IDs associated with the links got scraped. You need to keep them.

In case of Fate/zero, the link for prologue -> notes work, but the link that goes back to prologue is still broken.

**Comment by dteviot** [comment]
@toshiya44
I assume you mean the element with the id attribute matching the # part of the &lt;a&gt;'s href has been deleted.  (If I'm wrong, please correct me.)

I've just done https://www.baka-tsuki.org/project/index.php?title=Mushi_to_Medama for volume 1 and the following links (in translator notes) work.
- kakka
- Rinne
- Daddy-Long-Legs

However, "Mushi" does not.
Investigating further.  Looks like problem is util.removeUneededIds().  Which should really be called, removeUnusedIds().
It only checks for ids that are used within a page, not across pages.

I don't think this function is really needed. (If I recall correctly, it fixed some epubcheck warnings that were given for incorrectly formatted IDs, which I doubt gave any EPUB readers problems.)
So I've removed it from Experimental tab branch.  If necessary, will do revised version that works across all pages.

**Comment by dteviot** [comment]
@toshiya44
Apologies for nagging, but it's been 2 weeks with no response.
Did commit fix https://github.com/dteviot/WebToEpub/issues/98#issuecomment-354669165 for you?

**Comment by toshiya44** [comment]
Thank you for the ping @dteviot . I got distracted and didn't realize I haven't posted the feedback...

Yes, the issue in https://github.com/dteviot/WebToEpub/issues/98#issuecomment-354669165 is fixed. I also tested again Mushi to Medama and found no errors there either. Thank you for the fix :+1:

**Comment by dteviot** [comment]
I'm going to call this closed.
